diff --git a/Makefile b/Makefile
index 92ab23f..83d70c4 100644
--- a/Makefile
+++ b/Makefile
@@ -128,14 +128,19 @@ SUBDIRS+=lua-signal
 endif
 
 CHDKPTP_EXE=chdkptp$(EXE_EXTRA)$(EXE)
+CHDKPTP_LIB=chdkptp$(EXE_EXTRA)$(LIB)
 
 EXES=$(CHDKPTP_EXE)
+LIBS=$(CHDKPTP_LIB)
 
-all: $(EXES)
+all: $(EXES) $(LIBS)
 
 SRCS=properties.c ptp.c chdkptp.c lbuf.c liveimg.c rawimg.c luautil.c $(PTPIP_SRCS)
 OBJS=$(SRCS:.c=.o)
 
+$(CHDKPTP_LIB): $(OBJS)
+	$(CC) -shared -o $@ lfs/lfs.o $^ $(LDFLAGS)
+
 $(CHDKPTP_EXE): $(OBJS)
 	$(CC) -o $@ lfs/lfs.o $^ $(LDFLAGS)
 
diff --git a/chdkptp.c b/chdkptp.c
index de0662f..7535c3e 100644
--- a/chdkptp.c
+++ b/chdkptp.c
@@ -2197,7 +2197,11 @@ static const luaL_Reg lua_errlib[] = {
 };
 
 
-static int chdkptp_registerlibs(lua_State *L) {
+int luaopen_chdkptp(lua_State *L) {
+	usb_init();
+	luaopen_lfs(L);
+	luaopen_lbuf(L);
+	luaopen_rawimg(L);
 	/* set up meta table for error object */
 	luaL_newmetatable(L,CHDK_API_ERROR_META);
 	lua_pushcfunction(L,api_error_tostring);
@@ -2266,13 +2270,9 @@ int main(int argc, char ** argv)
 	g_argv = argv;
 	/* register signal handlers */
 //	signal(SIGINT, ptpcam_siginthandler);
-	usb_init();
 	lua_State *L = luaL_newstate();
 	luaL_openlibs(L);
-	luaopen_lfs(L);
-	luaopen_lbuf(L);
-	luaopen_rawimg(L);	
-	chdkptp_registerlibs(L);
+	luaopen_chdkptp(L);
 	int r=exec_lua_string(L,"require('main')");
 	uninit_gui_libs(L);
 	lua_close(L);
diff --git a/include.mk b/include.mk
index dbf51ca..2e537ae 100644
--- a/include.mk
+++ b/include.mk
@@ -9,6 +9,7 @@ ifeq ($(HOSTPLATFORM),MINGW)
 OSTYPE=Windows
 EXE=.exe
 DLL=.dll
+LIB=.dll
 # Note may be freetype or freetype6 depending on CD version, zlib requried for 5.5 and later
 CD_FREETYPE_LIB=freetype6 z
 #CD_FREETYPE_LIB=freetype z
@@ -17,6 +18,7 @@ ifeq ($(HOSTPLATFORM),Linux)
 OSTYPE=Linux
 EXE=
 DLL=.so
+LIB=.so
 CD_FREETYPE_LIB=freetype z
 endif
 ifeq ($(HOSTPLATFORM),Darwin)
@@ -24,6 +26,7 @@ OSTYPE=Darwin
 EXE=
 # TODO .dylib? only used for signal
 DLL=.so
+LIB=.so
 # TODO?
 CD_FREETYPE_LIB=freetype z
 endif
@@ -32,7 +35,7 @@ endif
 EXE_EXTRA=

 CC=gcc
-CFLAGS=-DCHDKPTP_OSTYPE=\"$(OSTYPE)\" -Wall
+CFLAGS=-DCHDKPTP_OSTYPE=\"$(OSTYPE)\" -Wall -fPIC -g
 LDFLAGS=
 #LD=ld
 
